node/test/lib/fake-egress-nodes.js

```diff
@@ -0,0 +1,91 @@
+// Copyright (c) 2015 Uber Technologies, Inc.
+
+// Permission is hereby granted, free of charge, to any person obtaining a copy
+// of this software and associated documentation files (the "Software"), to deal
+// in the Software without restriction, including without limitation the rights
+// to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
+// copies of the Software, and to permit persons to whom the Software is
+// furnished to do so, subject to the following conditions:
+//
+// The above copyright notice and this permission notice shall be included in
+// all copies or substantial portions of the Software.
+//
+// THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+// IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+// FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+// AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+// LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+// OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+// THE SOFTWARE.
+
+'use strict';
+
+var inherits = require('util').inherits;
+var EventEmitter = require('../../lib/event_emitter.js');
+
+module.exports = FakeEgressNodes;
+
+// e.g., topology = {'alice': ['127.0.0.1:4040'], 'bob': ['127.0.0.1:4041']}, '127.0.0.1:4040'
+function FakeEgressNodes(options) {
+    if (!(this instanceof FakeEgressNodes)) {
+        return new FakeEgressNodes(options);
+    }
+
+    var self = this;
+
+    EventEmitter.call(self);
+
+    self.hostPort = options.hostPort;
+
+    // Everyone must mutate this
+    self.topology = options.topology;
+    self.kValue = options.kValue;
+    self.relayChannels = options.relayChannels;
+
+    self.membershipChangedEvent = self.defineEvent('membershipChanged');
+}
+
+inherits(FakeEgressNodes, EventEmitter);
+
+FakeEgressNodes.prototype.isExitFor = function isExitFor(serviceName) {
+    var self = this;
+
+    return self.topology[serviceName].indexOf(self.hostPort) >= 0;
+};
+
+FakeEgressNodes.prototype.getRandomNodes =
+function getRandomNodes(serviceName) {
+    var self = this;
+
+    var hosts = [];
+
+    for (var i = 0; i < self.kValue; i++) {
```

Comment:
Not actually a good random sampler, but doesn't matter #justsayin.

How to do it right: select the top-k randomly chosen scores, e.g. this one-liner in Python:
```
[
    letter
    for score, letter in heapq.nlargest(10, (
        (random.random(), letter)
        for letter in string.ascii_letters))
]
```

I.e.:
- get you a min heap implementation
- iterate your input array, for each item
  - compute a random score
  - if heap is under k in size, push score and item
  - else if score is greater than head of heap, replace and sift down
- strip the scores off and return just the items out of your heap